web-admin/src/features/shared/refetch-interval.ts

```diff
@@ -0,0 +1,79 @@
+import type { Query } from "@tanstack/svelte-query";
+import type {
+  V1ListResourcesResponse,
+  V1Resource,
+} from "@rilldata/web-common/runtime-client";
+import type { HTTPError } from "@rilldata/web-common/runtime-client/fetchWrapper";
+import { writable, type Writable } from "svelte/store";
+
+export const INITIAL_REFETCH_INTERVAL = 200; // Start at 200ms for immediate feedback
+export const MAX_REFETCH_INTERVAL = 2_000; // Cap at 2s
+export const BACKOFF_FACTOR = 1.5;
+
+export function isResourceErrored(resource: V1Resource) {
+  return !!resource?.meta?.reconcileError;
+}
+
+export function isResourceReconciling(resource: V1Resource) {
+  return (
+    resource?.meta?.reconcileStatus === "RECONCILE_STATUS_PENDING" ||
+    resource?.meta?.reconcileStatus === "RECONCILE_STATUS_RUNNING"
+  );
+}
+
+export function pollUntilResourcesReconciled(
+  currentInterval: number,
+  data: V1ListResourcesResponse | undefined,
+  query: Query<V1ListResourcesResponse, HTTPError>,
+): number | false {
+  if (query.state.error) return false;
+  if (!data?.resources) return false; // Stop polling if no resources
+
+  const hasErrors = data.resources.some(isResourceErrored);
+  const hasReconcilingResources = data.resources.some(isResourceReconciling);
+
+  if (hasErrors || !hasReconcilingResources) {
+    return false;
+  }
+
+  return Math.min(currentInterval * BACKOFF_FACTOR, MAX_REFETCH_INTERVAL);
+}
+
+function createRefetchIntervalStore() {
+  const store: Writable<number> = writable(INITIAL_REFETCH_INTERVAL);
```

Comment:
> One idea, could we add custom state to either query.meta or query.state, e.g. query.state.currentRefetchInterval = X?

I'm curious, did you consider this^ idea?

Reply:
removed the store code and used the query's state directly https://github.com/rilldata/rill/pull/7216/commits/7c1017755ef048fd2b9ce6bb0adffcf5e8421ae0